pydantic/main.py

```diff
@@ -42,6 +43,10 @@
 # the `BaseModel` class, since that's defined immediately after the metaclass.
 _base_class_defined = False
 
+_generic_types_cache: _utils.LimitedDict[
+    typing.Tuple[type[Any], Any, typing.Tuple[Any, ...]], type[BaseModel]
```

Comment:
It might be worth looking at #3829 #4551 to check we don't introduce any of those problems again - the memory leak issue should be solved by `LimitedDict`, but the `TypeError` could still exist I guess.